Home 7

Y-Pazniak opened the pull request:
Здравствуйте, Максим. 
Спасибо за лекцию о строках - было очень познавательно. 
Отправляю домашнее задание для проверки.

Y-Pazniak committed:
Test commit and Push

Y-Pazniak committed:
Delete Home_7/TestString/src directory

Y-Pazniak committed:
Commit for 7 homework

Y-Pazniak commented on Home_7/Task_7_1/src/Main.java:
```diff
@@ -0,0 +1,7 @@
+public class Main {
```
Могу ли я тут писать?

Y-Pazniak committed:
Commit for 7 homework

Y-Pazniak committed:
Wrong comment at 36 line fixed

Y-Pazniak committed:
Merge remote-tracking branch 'origin/Home_7' into Home_7

MaximPilipeiko commented on Home_7/Task_7_3/src/Main.java:
```diff
@@ -0,0 +1,44 @@
+        //String[] words = s.split("\\W+"); //как оказалось, эта регулярка не дружит с кириллицей - а жаль :(
+        String[] words = s.split("[,:;!?\"\\s+]"); //здесь тоже попытался в регулярные выражения - убрал запятые, точки с запятой, другие мешающие знаки
+        for (String word : words) {
+            if ((word.length() > 1) && (word.equalsIgnoreCase((new StringBuilder(word).reverse()).toString()))) { //сравнил: если слово длиннее 1 символа и перевернутое равно само себе - увеличиваем переменную на 1
```
Хорошо бы вынести в отдельный метод

Y-Pazniak replied:
Сделали: https://github.com/Y-Pazniak/AN11_Android/commit/c7b25288247cf18f02fb3adff9d9ba1234f049ec
(если правильно понял, что именно нужно вынести в отдельный метод, если неправильно - пожалуйста, сообщите)

MaximPilipeiko commented on Home_7/Task_7_4/src/Main.java:
```diff
@@ -0,0 +1,43 @@
+    public static int findDifferentLetterCount(final String s) {  //этот метод считает, сколько разных символов в переданном слове
+        int local = 1; //если мы получили слово, то как минимум 1 символ в слове уже есть
+        for (int i = 0; i < s.length() - 1; i++) {
+            if (!s.substring(i + 1).contains(String.valueOf(s.charAt(i)))) { //отсекаем проверяемый символв и символы перед ним и смотрим, есть ли такой символ - если больше его нет, а строка продолжается, увеличиваем к-во символов
```
Интересный подход. Только породили много объектов трок.

Y-Pazniak replied:
Кажется, сейчас получилось лучше: https://github.com/Y-Pazniak/AN11_Android/pull/3/commits/69ce1467649d3d98ced869ce473199110f6ef598

MaximPilipeiko replied:
Определенно лучше. Спасибо за ссылку на комит.

MaximPilipeiko commented on Home_7/Task_7_5/src/Util.java:
```diff
@@ -0,0 +1,10 @@
+    public static String createMessage(final Patient patient, final boolean covidTest, final int antibodiesCount) {
+        String aboutTest = covidTest ? ", unfortunately test is POSITIVE. Stay at home at 7 days. " : ", fortunately, test is NEGATIVE. Enjoy your regular life. ";
+        String aboutAntibodies = (antibodiesCount > 0) ? "Your antibodies count is: " + antibodiesCount : "You have no antibodies.";
+        return "Dear " + //я сначала хотел сделать через StringBuilder, но Idea посоветовала просто стринг вернуть - мне кажется, она права, так лучше выглядит
```
Эх, а я хотел именно StringBuilder увидеть. Но так конечно тоже хорошо.

Y-Pazniak replied:
Если мы проходили StringBuilder, логично, что нужно было использовать его. Просто я поленился писать ветвления, моя вина. Здесь исправился: https://github.com/Y-Pazniak/AN11_Android/commit/6d555a52a9308a89774fbb197ab98f4cf88f7563

Y-Pazniak replied:
Вот тут немножко подхорошил: https://github.com/Y-Pazniak/AN11_Android/commit/5690df6c78e7d9b95fe806c819843b403428b59b

MaximPilipeiko replied:
именно так и хотел увидеть, спасибо

MaximPilipeiko commented on Home_7/Task_7_6/src/Main.java:
```diff
@@ -0,0 +1,42 @@
+public class Main { //"Дан текст, отсортировать слова в тексте по их количеству" - немножко неоднозначное задание, я понял его как "отсортировать текст по длине слов в нем", если неправильно - напишите, я переделаю
```
Согласен что не однозначно, имелось ввиду отсортировать слова по количеству вхождений этого слово в текст

Y-Pazniak replied:
Кажется, переделал. Действительно, по вхождениям слов в текст без листов и мапы сделать существенно труднее.
https://github.com/Y-Pazniak/AN11_Android/commit/535bb03784935a91863e25fe58e42dd6906bd34e

MaximPilipeiko replied:
Отличный алгоритм, зря недоволен.

MaximPilipeiko reviewed (approved):
Классная работа!
Если получится попробуй переделать 4

Y-Pazniak committed:
"Хорошо бы вынести в отдельный метод"
Вынес проверку слова и слова задом наперед в отдельный метод (45-46 строки)

Y-Pazniak committed:
"Хорошо бы вынести в отдельный метод"
Вынес проверку слова и слова задом наперед в отдельный метод (45-46 строки)

Y-Pazniak committed:
"Интересный подход. Только породили много объектов трок."
Исправили: сократили до объекта - массива чаров, в котором считаем отличающиеся символы

Y-Pazniak committed:
Merge remote-tracking branch 'origin/Home_7' into Home_7

Y-Pazniak committed:
Исправлен класс Util на использование StringBuilder

Y-Pazniak committed:
Исправлен класс Util на использование StringBuilder

Y-Pazniak committed:
Максим, отправляю программу, которая сортирует текст по убыванию количества слов в нем.
Прошу прощения за громоздкий код: я понимаю, что он чудовищный, но по-другому не придумал как сделать. Уверен, есть более легкий алгоритм, но до него не догадался.
Вариант через мапу вы попросили не реализовывать, другие Листы я так понимаю тоже не надо использовать.
Вроде бы программа работает, но я ей не очень доволен.
